### PR TITLE
exclude `DestinationKeyspaceID,DestinationKeyspaceIDs` from the plan cache key prefix

### PR DESCRIPTION
**Pull request by inexplicable**
as this is causing our plan cache to quickly reach capacity at pinterest.
we does use `DestinationKeyspaceID` heavily, and the `keyspaceID` has a really large cardinality, causing frequent evictions.

[update] simply excluding `destination.String()` doesn't work, it could cause wrong destination cached in the plan. as a result, added resolved shard as part of the cache key.

### CODE REVIEW

**Comment by dweitzman** [comment]
In recent versions of vitess we see vtgate plan cache keys that look like `<keyspace>@<tableType>DestinationKeyspaceID(99298cd3d80e95ec):<query>`. This may be related to v2 query planning being replaced by the v3 stuff

**Comment by systay** [comment]
Hi there!

Sorry for not answering faster. The changes look good to me - the only thing I would love to have is a test or two that test the interaction with the plan cache. I think it would make it much easier to understand what the external behaviour we are after is.

File: go/vt/vtgate/vcursor_impl.go [hunk 1]
```diff
@@ -420,6 +422,20 @@ func parseDestinationTarget(targetString string, vschema *vindexes.VSchema) (str
...
+		switch vc.destination.(type) {
+		case key.DestinationKeyspaceID, key.DestinationKeyspaceIDs:
+			resolved, _, err := vc.ResolveDestinations(vc.keyspace, nil, []key.Destination{vc.destination})
+			if err == nil && len(resolved) > 0 {
+				shards := make([]string, len(resolved))
+				for i := 0; i < len(shards); i++ {
+					shards[i] = resolved[i].Target.GetShard()
+				}
+				sort.Strings(shards)
+				return fmt.Sprintf("%s%sDestinationKeyspaceIDsResolved(%s)", vc.keyspace, vindexes.TabletTypeSuffix[vc.tabletType], strings.Join(shards, ","))
```
Comment (harshit-gangal) [inline go/vt/vtgate/vcursor_impl.go:434]: we do not need to be descriptive in plan key. DestinationKeyspaceIDsResolved(%s) can be just %s

Comment (inexplicable) [reply go/vt/vtgate/vcursor_impl.go:434]: got it, agree that the verbosity might take extra memory, will shorten it so the overhead could be reduced.

**Comment by inexplicable** [comment]
> Hi there!
> 
> Sorry for not answering faster. The changes look good to me - the only thing I would love to add would be a test or two that test the interaction with the plan cache. I think it would make it much easier to understand what the external behaviour we are after is. Let me know if you need a hand writing them.

make sense, i've updated the `executor_test` to add the `plancachekey` test case

**Pull request merged**